sockittome.js

```diff
@@ -241,63 +241,79 @@ var rooms = {};
 
 // Redux Reducers
 
-let lastId = 0;
-let songIndex = 0;
-
 const CREATE_ROOM = "createRoom";
 const SONG_ADDED = "songAdded";
 const PLAY_PRESSED = "playPressed";
 const PAUSE_PRESSED = "pausePressed";
 const PREV_PRESSED = "prevPressed";
 const NEXT_PRESSED = "nextPressed";
 
-export default function playlistReducer(state = [], action) {
+let nextRoomId = 0;
+let songIndex = 0;
+
+export default function playlistReducer(state = {}, action) {
   switch (action.type) {
-    case "CREATE_ROOM":
-      return state.concat([
-        {
-          roomId: ++lastId,
+    case "CREATE_ROOM": {
+      const roomId = ++nextRoomId;
+      return {
+        ...state,
+        [roomId]: {
+          roomId,
           socketUserMap: action.payload.user,
-          playlist: [],
-          queue: [],
+          playlist: ["https://cdn.glitch.me/aa3b905c-b152-45c7-9d6f-45c998461107%2Fkaraoke_nite_welcome_video_hd.mp4"],
+          queue: ["https://cdn.glitch.me/aa3b905c-b152-45c7-9d6f-45c998461107%2Fkaraoke_nite_welcome_video_hd.mp4"],
           currentPlayingIndex: 0,
           currentPlayingTime: 0,
           isPlaying: false,
         }
-      ]);
+      }
+    }
     case "SONG_ADDED": {
-      return state.map(room => room.roomId === action.payload.id 
-        ? { 
-          ...room, 
-          playlist: [...room.playlist, action.payload.song],
-          queue: [...room.queue, action.payload.song] 
+      return {
+        ...state,
+        [action.payload.id]: {
+          ...state[action.payload.id],
+          playlist: [...state[action.payload.id].playlist, action.payload.song],
+          queue: [...state[action.payload.id].queue, action.payload.song] 
         }
-        : room
-      )
+      }
     }
     case "PLAY_PRESSED": {
-      return state.map(room => room.roomId === action.payload.id 
-        ? { ...room, isPlaying: true }
-        : room
-      )
+      return {
+        ...state,
+        [action.payload.id]: {
+          ...state[action.payload.id],
+          isPlaying: true
+        }
+      }
     }
     case "PAUSE_PRESSED": {
-      return state.map(room => room.roomId === action.payload.id 
-        ? { ...room, isPlaying: false }
-        : room
-      )
+      return {
+        ...state,
+        [action.payload.id]: {
+          ...state[action.payload.id],
+          isPlaying: false
+        }
+      }
     }
     case "PREV_PRESSED": {
-      return state.map(room => room.roomId === action.payload.id 
-        ? { ...room, currentPlayingIndex: --songIndex }
-        : room
-      )
+      return {
+        ...state,
+        [action.payload.id]: {
+          ...state[action.payload.id],
+          currentPlayingIndex: --songIndex
```

Comment:
```tsx

    case "PREV_PRESSED": {
     const roomId = action.payload.id;
     const room = state[roomId];
      return {
        ...state,
        [roomId]: {
          ...room,
          currentPlayingIndex: room.currentPlayingIndex - 1 ?? 0, // Don't go below 0
       }
    }
}

Reply:
The reducer pattern's main benefit is for computing the next state from both the input and the current state.

Actions can have empty payloads in certain cases because the type of action only requires the previous state to do the specific task.

```tsx
case 'PLAY': {
   return {
      ...state,
      isPlaying: true
    }
}
case 'PAUSE': {
   return {
      ...state,
      isPlaying: false
    }
}
```

Reply:
just updated `PREV_PRESSED`. for this one, shouldn't it be:

```js
case "PLAY_PRESSED": {
      const roomId = action.payload.id;
      const room = state[roomId];
      return {
        ...state,
        [roomId]: {
          ...room,
          isPlaying: true
        }
      }
    }
    ```

Reply:
yep!